packages/console/src/components/CodeEditor/index.tsx

```diff
@@ -12,11 +12,39 @@ type Props = {
   isReadonly?: boolean;
   value?: string;
   onChange?: (value: string) => void;
+  tabSize?: number;
 };
 
-const CodeEditor = ({ className, language, isReadonly = false, value = '', onChange }: Props) => {
+const CodeEditor = ({
+  className,
+  language,
+  isReadonly = false,
+  value = '',
+  onChange,
+  tabSize = 4,
+}: Props) => {
+  const textareaRef = useRef<HTMLTextAreaElement>(null);
+
   const handleChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
-    onChange?.(event.currentTarget.value);
+    const { value } = event.currentTarget;
+    onChange?.(value);
+  };
+
+  const handleKeydown = (event: KeyboardEvent<HTMLTextAreaElement>) => {
+    if (event.key === 'Tab') {
+      const { value, selectionStart } = event.currentTarget;
+
+      event.preventDefault();
+      const newText =
+        value.slice(0, selectionStart) + ' '.repeat(tabSize) + value.slice(selectionStart);
+
+      // Need to update value to set selection without useEffect
+      // eslint-disable-next-line @silverhand/fp/no-mutation
+      event.currentTarget.value = newText;
```

Comment:
@charIeszhao Because `setSelectionRange` has to be called after the state is changed, I force to set the value of textarea.

I there a better way to do it?

Reply:
Call `onChange(newText)` directly?

Reply:
https://erikmartinjordan.com/react-textarea-indent

Reply:
The status quo looks also good to me

Reply:
Yes, thanks @gao-sun.
However, after discussion we think the current implementation is acceptable as other alternatives will either require `useEffect` and extra state/ref in order to make it work. So this one right now looks elegant enough to proceed.